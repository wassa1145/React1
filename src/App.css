body {
  background-color: #f8f8f8;
  font-family: Arial, Helvetica, sans-serif;
}
.page {
  max-width: 1440px;
  min-width: 320px;
  margin: 0 auto;
  box-sizing: border-box;
}
.content {
  min-width: 445px;
  max-width: 600px;
  margin: 0 auto;
  background-color: #e4ebf8;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  height: 100vh;
}
