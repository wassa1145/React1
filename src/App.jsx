import React, { useState, useEffect, useRef } from 'react';
import './App.css';
import { Form } from './components/FormFunc/Form';
import { List } from './components/List/List';
import { CONSTANTS } from './constants';

export const App = () => {
  const [messageList, setMessageList] = useState([]);
  const listEl = useRef(null);
  const addMessage = (message) => {
    setMessageList([
      ...messageList,
      {
        author: CONSTANTS.USER,
        message,
      },
    ]);
  };
  const scrollList = () => {
    listEl.current.scrollTop = listEl.current.scrollHeight;
  };

  useEffect(() => {
    if (
      messageList.length &&
      !messageList[messageList.length - 1].systemMessage
    ) {
      const timeout = setTimeout(() => {
        setMessageList([
          ...messageList,
          {
            message: CONSTANTS.DEFAULT_MESSAGE,
            author: CONSTANTS.BOT,
            systemMessage: true,
          },
        ]);
      }, 1000);
      return () => {
        clearTimeout(timeout);
      };
    }
    scrollList();
  }, [messageList.length]);

  return (
    <div className="content">
      <List messages={messageList} refEl={listEl} />
      <Form addMessage={addMessage} />
    </div>
  );
};
