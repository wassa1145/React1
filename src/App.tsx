import React, { FC, useState, useMemo } from 'react';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { Provider } from 'react-redux';
import { ThemeContext, defaultContext } from './utils/ThemeContext';
import { nanoid } from 'nanoid';
import './App.css';
import { store } from './store';
import { ChatList } from './components/ChatList/ChatList';
import { CONSTANTS } from './constants';
import { Header } from './components/Header/Header';
import { Home } from './pages/Home';
import { Profile } from './pages/Profile';
import { Chats } from './pages/Chats';

export interface Chat {
  id: string;
  name: string;
}

interface Message {
  id: string;
  author: string;
  message: string;
  systemMessage?: boolean;
}

export interface Messages {
  [key: string]: Message[];
}

const initialMessage: Messages = {
  default: [
    {
      id: '1',
      author: CONSTANTS.USER,
      message: 'Hello geekbrains',
    },
  ],
};

export const App: FC = () => {
  const [messages, setMessages] = useState<Messages>(initialMessage);
  const [theme, setTheme] = useState(defaultContext.theme);

  const chatList = useMemo(
    () =>
      Object.entries(messages).map((chat) => ({
        id: nanoid(),
        name: chat[0],
      })),
    [Object.entries(messages).length]
  );

  const onAddChat = (chat: Chat) => {
    setMessages({
      ...messages,
      [chat.name]: [],
    });
  };
  const deleteChat = (chatName: string) => () => {
    const newMessages = { ...messages };
    delete newMessages[chatName];
    setMessages(newMessages);
  };

  const toggleTheme = () => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  };

  return (
    <Provider store={store}>
      <ThemeContext.Provider
        value={{
          theme,
          toggleTheme,
        }}
      >
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<Header />}>
              <Route index element={<Home />} />
              <Route path="profile" element={<Profile />} />
              <Route path="chats">
                <Route
                  index
                  element={
                    <ChatList
                      chats={chatList}
                      onAddChat={onAddChat}
                      deleteChat={deleteChat}
                    />
                  }
                />
                <Route
                  path=":chatId"
                  element={
                    <Chats
                      messages={messages}
                      setMessages={setMessages}
                      chatList={chatList}
                      onAddChat={onAddChat}
                      deleteChat={deleteChat}
                    />
                  }
                />
              </Route>
            </Route>
            <Route path="*" element={<h2>404</h2>} />
          </Routes>
        </BrowserRouter>
      </ThemeContext.Provider>
    </Provider>
  );
};
