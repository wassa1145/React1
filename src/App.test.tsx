import React from 'react';
import { App } from './App';
import { act, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';

beforeEach(() => {
  render(
    <BrowserRouter>
      <App />
    </BrowserRouter>
  );
});

describe('App', () => {
  it('render with snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('render header', () => {
    expect(screen.getAllByText(/Home/)[0]).toBeInTheDocument();
    expect(screen.getByText(/Chats/)).toBeInTheDocument();
    expect(screen.getAllByText(/Profile/)[0]).toBeInTheDocument();
  });

  it('navigates home when you click Home', () => {
    act(() => {
      const goHomeLink = document.querySelector('#home');
      goHomeLink?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    expect(screen.getByText(/Home Page/)).toBeInTheDocument();
  });

  it('navigates chats when you click Chats', () => {
    act(() => {
      const goChatsLink = document.querySelector('#Chats');
      goChatsLink?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    expect(screen.getByText(/Список чатов/)).toBeInTheDocument();
  });

  it('delete chat', async () => {
    act(() => {
      const goChatsLink = document.querySelector('#Chats');
      goChatsLink?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const countChat = screen.queryAllByRole('chat').length;
    await userEvent.click(screen.getByRole('button-delete'));
    expect(screen.queryAllByRole('chat').length).toBe(countChat - 1);
  });

  // Падает с ошибкой  Unable to perform pointer interaction as the element has or inherits pointer-events set to "none".
  // не понимаю почему.

  // it('add chat', async () => {
  //   act(() => {
  //     const goChatsLink = document.querySelector('#Chats');
  //     goChatsLink?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  //   });
  //   await userEvent.type(screen.getByRole('input-chat-name'), 'new chat');
  //   await userEvent.click(screen.getByText(/Добавить новый чат/));
  //   expect(screen.getByText(/new chat/)).toBeInTheDocument();
  // });

  // it('clear textarea on submit', async () => {
  //   await userEvent.type(screen.getByRole('textbox'), 'new message');
  //   await userEvent.click(screen.getByTestId('submit-button'));
  //   expect(screen.getByRole('textbox')).toHaveValue('');
  // });

  // it('add message on submit', async () => {
  //   const countMessages = screen.queryAllByRole('message').length;
  //   await userEvent.type(screen.getByRole('textbox'), 'new message');
  //   await userEvent.click(screen.getByTestId('submit-button'));
  //   expect(screen.queryAllByRole('message').length).toBe(countMessages + 1);
  // });

  // it('add system message', async () => {
  //   const countMessages = screen.queryAllByRole('message').length;
  //   await userEvent.type(screen.getByRole('textbox'), 'new message');
  //   await userEvent.click(screen.getByTestId('submit-button'));
  //   await waitFor(
  //     () =>
  //       expect(screen.queryAllByRole('message').length).toBe(countMessages + 2),
  //     {
  //       timeout: 1500,
  //     }
  //   );
  // });
});
