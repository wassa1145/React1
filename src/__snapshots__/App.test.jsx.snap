// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render with snapshot 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <div
      class="list"
      data-testid="messages-list"
    >
      <ul />
    </div>
    <form
      class="form"
      data-testid="form"
    >
      <textarea
        class="form__textarea"
        data-testid="textarea"
        placeholder="Введите сообщение"
        rows="6"
      />
      <button
        class="form__button"
        data-testid="submit-button"
        disabled=""
        type="submit"
      >
        Отправить
      </button>
    </form>
  </div>
</DocumentFragment>
`;
